File: scripts/file-issues.py
#!/usr/bin/env python3
"""Parse docs/code-quality-issues.md and create GitHub issues with `gh`."""
import os
import re
import subprocess
import sys
import tempfile
from pathlib import Path

REPO = os.environ.get("REPO", "faketut/WeAgree")
DRY = os.environ.get("DRY") == "1"
FILE = Path(sys.argv[1] if len(sys.argv) > 1 else "docs/code-quality-issues.md")

text = FILE.read_text()
# Split into sections beginning with "## N. Title"
sections = re.split(r"\n(?=## \d+\.\s)", text)
sections = [s for s in sections if re.match(r"^## \d+\.\s", s)]

tmpdir = Path(tempfile.mkdtemp(prefix="wa-issues-"))
ok = 0
fail = 0

for idx, sec in enumerate(sections, 1):
    lines = sec.split("\n")
    m = re.match(r"^## \d+\.\s+(.+?)\s*$", lines[0])
    if not m:
        continue
    title = m.group(1).strip()
    body = "\n".join(lines[1:]).strip()
    body = re.sub(r"\n+---\s*$", "", body).strip()

    lbl_m = re.search(r"^\*\*Labels:\*\*\s*(.+)$", body, re.M)
    labels = []
    if lbl_m:
        labels = [s.replace("`", "").strip() for s in lbl_m.group(1).split(",")]
        labels = [l for l in labels if l]

    body_file = tmpdir / f"{idx}.md"
    body_file.write_text(body)

    cmd = ["gh", "issue", "create", "-R", REPO, "-t", title, "-F", str(body_file)]
    for l in labels:
        cmd += ["-l", l]

    if DRY:
        print(f"[dry] {idx:>2}. {title}  labels={labels}")
        ok += 1
        continue

    res = subprocess.run(cmd, capture_output=True, text=True)
    if res.returncode != 0:
        print(f"FAIL #{idx} {title}: {res.stderr.strip() or res.stdout.strip()}")
        fail += 1
        continue
    url = res.stdout.strip().splitlines()[-1]
    print(f"OK  #{idx:>2} {title} -> {url}")
    ok += 1

print(f"\n{ok} ok, {fail} failed, {len(sections)} total.")

File: scripts/file-issues.mjs
#!/usr/bin/env node
// Parses docs/code-quality-issues.md and creates one GitHub issue per `## N. Title` section.
// Labels come from the `**Labels:**` line inside each section.
import fs from "node:fs";
import { spawnSync } from "node:child_process";
import path from "node:path";

const REPO = process.env.REPO || "faketut/WeAgree";
const DRY = process.env.DRY === "1";
const FILE = process.argv[2] || "docs/code-quality-issues.md";

const md = fs.readFileSync(FILE, "utf8");
// Split on top-level "## " headings, drop preamble (before first "## 1.")
const parts = md.split(/\n(?=## \d+\.\s)/g).filter((s) => /^## \d+\.\s/.test(s));

const tmpDir = fs.mkdtempSync("/tmp/wa-issues-");
const created = [];

for (const section of parts) {
  const lines = section.split("\n");
  const headingMatch = lines[0].match(/^## \d+\.\s+(.+?)\s*$/);
  if (!headingMatch) continue;
  const title = headingMatch[1].trim();

  // Body = everything after the heading line, trimmed; cut trailing "---" separator block if present.
  let body = lines.slice(1).join("\n").trim();
  body = body.replace(/\n+---\s*$/m, "").trim();

  // Extract labels
  const labelLine = body.match(/^\*\*Labels:\*\*\s*(.+)$/m);
  const labels = labelLine
    ? labelLine[1]
        .split(",")
        .map((s) => s.replace(/`/g, "").trim())
        .filter(Boolean)
    : [];

  const bodyFile = path.join(tmpDir, `${created.length + 1}.md`);
  fs.writeFileSync(bodyFile, body);

  const args = [
    "issue",
    "create",
    "-R",
    REPO,
    "-t",
    title,
    "-F",
    bodyFile,
  ];
  for (const l of labels) {
    args.push("-l", l);
  }

  if (DRY) {
    console.log(`[dry] gh ${args.map((a) => (a.includes(" ") ? JSON.stringify(a) : a)).join(" ")}`);
    created.push({ title, labels, url: null });
    continue;
  }

  const res = spawnSync("gh", args, { encoding: "utf8" });
  const out = (res.stdout || "").trim();
  const err = (res.stderr || "").trim();
  if (res.status !== 0) {
    console.error(`FAIL: ${title}\n${err || out}`);
    continue;
  }
  const url = out.split("\n").pop();
  console.log(`OK  #${created.length + 1}: ${title} -> ${url}`);
  created.push({ title, labels, url });
}

console.log(`\nDone. ${created.filter((c) => c.url || DRY).length}/${parts.length} issues processed.`);
